File: medium/Array/product-of-array-except-self/product_except_self.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> res(n, 1);

        // Prefix
        int prefix = 1;
        for (int i = 0; i < n; ++i) {
            res[i] = prefix;
            prefix *= nums[i];
        }

        // Suffix
        int suffix = 1;
        for (int i = n-1; i >= 0; --i) {
            res[i] *= suffix;
            suffix *= nums[i];
        }

        return res;
    }
};

// Test
#include <iostream>
int main() {
    Solution sol;
    vector<int> v = {1,2,3,4};
    auto ans = sol.productExceptSelf(v);
    for (int x : ans) cout << x << " ";
    return 0;
}

File: medium/Array/product-of-array-except-self/product_except_self.py
from typing import List

def productExceptSelf(nums: List[int]) -> List[int]:
    """
    Return an array where each element is the product of all elements except itself.
    Approach: Prefix & Suffix Product
    Time: O(n)
    Space: O(1) (output array does not count)
    """
    n = len(nums)
    res = [1] * n

    # Prefix (left product)
    prefix = 1
    for i in range(n):
        res[i] = prefix
        prefix *= nums[i]

    # Suffix (right product)
    suffix = 1
    for i in range(n-1, -1, -1):
        res[i] *= suffix
        suffix *= nums[i]

    return res

if __name__ == "__main__":
    print(productExceptSelf([1,2,3,4]))  # [24,12,8,6]
